llvm/lib/SYCLLowerIR/SYCLOptimizeBarriers.cpp

```diff
@@ -0,0 +1,704 @@
+
+    if (hasFencedAccesses(BB, Required, StartInst, nullptr))
+      return false;
+
+    for (BasicBlock *Succ : successors(BB))
+      if (Visited.insert(Succ).second)
+        Worklist.emplace_back(Succ, nullptr);
+  }
+
+  return true;
+}
+
+// The back-to-back elimination on one BB.
+static bool eliminateBackToBackInBB(BasicBlock *BB,
+                                    SmallVectorImpl<BarrierDesc> &Barriers,
+                                    BBMemInfoMap &BBMemInfo) {
+  SmallVector<BarrierDesc, 8> Survivors;
+  bool Changed = false;
+  RegionMemScope BlockScope = BB ? BBMemInfo[BB] : RegionMemScope::Unknown;
+
+  // If there are no memory accesses requiring synchronization in this block,
+  // collapse all barriers to the single largest one.
+  if (BlockScope == RegionMemScope::None) {
+    bool HasUnknown = llvm::any_of(Barriers, [](const BarrierDesc &BD) {
+      return BD.ExecScope == Scope::Unknown || BD.MemScope == Scope::Unknown;
+    });
+    if (!HasUnknown) {
+      // Pick the barrier with the widest scope.
+      auto Best = std::max_element(
+          Barriers.begin(), Barriers.end(),
+          [](const BarrierDesc &A, const BarrierDesc &B) {
+            auto CmpExec = compareScopesWithWeights(B.ExecScope, A.ExecScope);
+            auto CmpMem = compareScopesWithWeights(B.MemScope, A.MemScope);
+            return (CmpExec == CompareRes::BIGGER ||
+                    (CmpExec == CompareRes::EQUAL &&
+                     CmpMem == CompareRes::BIGGER)) ||
+                   (CmpMem == CompareRes::BIGGER);
+          });
+
+      // Remove all other barriers in the block.
+      llvm::erase_if(Barriers, [&](BarrierDesc &BD) {
+        if (&BD == &*Best)
+          return false;
+        Changed |= eraseBarrierWithITT(BD);
+        return true;
+      });
+      return Changed;
+    }
+  }
+
+  // Otherwise do a sliding window compare of each barrier against the
+  // last survivor.
+  for (auto &Cur : Barriers) {
+    if (!Cur.CI)
+      continue; // already removed
+    while (!Survivors.empty()) {
+      BarrierDesc &Last = Survivors.back();
+      // Must share semantics to guess.
+      // TODO: actually allow semantics missmatch for barriers removal for
+      // several cases.
+      if (Last.Semantic != Cur.Semantic)
+        break;
+
+      auto CmpExec = compareScopesWithWeights(Last.ExecScope, Cur.ExecScope);
+      auto CmpMem = compareScopesWithWeights(Last.MemScope, Cur.MemScope);
+      RegionMemScope FenceLast = getBarrierFencedScope(Last.MemScope);
+      RegionMemScope FenceCur = getBarrierFencedScope(Cur.MemScope);
+
+      if (CmpExec == CompareRes::UNKNOWN || CmpMem == CompareRes::UNKNOWN ||
+          FenceLast == RegionMemScope::Unknown ||
+          FenceCur == RegionMemScope::Unknown)
+        break;
+
+      // If identical then drop Cur.
+      if (CmpExec == CompareRes::EQUAL && CmpMem == CompareRes::EQUAL) {
+        if (noFencedMemAccessesBetween(Last.CI, Cur.CI, FenceLast, BBMemInfo)) {
```

Comment:
just a note: there could be repeated classifyMemScope calculation of somes instructions in noFencedMemAccessesBetween, e.g. following case:
```
barrier(CrossDevice)
Instruction Set 1 (RegionMemScope == None)
barrier(Device)
Instruction Set 2 (RegionMemScope == None)
barrier(Workgroup)
Instruction Set 3 (RegionMemScope == None)
barrier(Subgroup)
```
I guess the case is rare, so probably no need to optimize.

Reply:
In general there are other ways to define fenced regions between barriers, but I haven't though about them until last Monday, when I found a similar work :) Re-making scanning and re-defining fenced regions is a possible enhancement for the pass.

Reply:
> In general there are other ways to define fenced regions between barriers, but I haven't though about them until last Monday, when I found a similar work :) Re-making scanning and re-defining fenced regions is a possible enhancement for the pass.

Sounding interesting, is there a link to the work?

Reply:
@wenju-he I meant CPU middle end pass, which while is not doing the same a this pass, yet have quite interesting idea for function preparation :)

Reply:
I see. Right it is not the same. I agree that a region based algorithm would be better. Basically the pass here is merging equivalent regions.